index.js

```diff
@@ -1,60 +1,67 @@
 'use strict';
 const path = require('path');
 const {constants: fsConstants} = require('fs');
-const {Buffer} = require('safe-buffer');
+const pEvent = require('p-event');
 const CpFileError = require('./cp-file-error');
 const fs = require('./fs');
 const ProgressEmitter = require('./progress-emitter');
 
+const cpFileAsync = async (source, destination, options, progressEmitter) => {
+	let readError;
+	const stat = await fs.stat(source);
+	progressEmitter.size = stat.size;
+
+	const read = await fs.createReadStream(source);
+	await fs.makeDir(path.dirname(destination));
+	const write = fs.createWriteStream(destination, {flags: options.overwrite ? 'w' : 'wx'});
+	read.on('data', () => {
+		progressEmitter.written = write.bytesWritten;
+	});
+	read.once('error', error => {
+		readError = new CpFileError(`Cannot read from \`${source}\`: ${error.message}`, error);
+		write.end();
+	});
+
+	let updateStats = false;
+	try {
+		const writePromise = pEvent(write, 'close');
```

Comment:
I'm not sure if it is strictly necessary to initialize the event watchers before `read.pipe(write)`.  The previous code added event watchers before piping so I kept it the same.  I'm also not positive that it's impossible for the `write` stream to emit 'close' or 'error' before next tick.

Reply:
> I'm also not positive that it's impossible for the write stream to emit 'close' or 'error' before next tick.

I've had problems in the past where they do, so always smart to initialize first.